### DISCUSSION TITLE
G7 Support?

### DISCUSSION THREAD

**Discussion by Miquest** [NightscoutFoundation/xDrip#2353] [Ideas]
The Dexcom G7 is out since a few weeks/months (I don't know it exactly...).
Will xDrip have a G7 Support in near future?
I would be happy if you implement it, because my Smartphone is not compatible with the G7 App and I want to upgrade from my G6.

Kind regards 
Thorin E.

**Comment by Navid200** [comment]
>  I would be happy if you implement it, 

Great to know that.
Can you implement it and open a PR?

If not, can you provide an expired device?

**Reply by Miquest** [reply]
Sorry, I'm no Java Developer...
I think Python and JavaScript are not the best programming languages to implement that... :\
If my t:slim has the Upgrade for the G7, sure!
I will work with AAPS and jwoglom's t:slim protocol until the implementation is done.

**Reply by steve8x8** [reply]
Estimating the chances that there will be a G7 BYODA (BYODG7A or whatever) - that might provide a first way to get G7 data into xDrip

(https://www.reddit.com/r/dexcom/comments/wgmg4c/comment/iplmrfi/?utm_source=reddit&utm_medium=web2x&context=3 seems to be the place where information comes directly from the horse's mouth)

**Reply by andyrozman** [reply]
G7 support in Tandem won't happen very soon (at least that would be my estimation), and even then it would be useless for AAPS. AAPS Tandem driver won't read G7(or G6) data from pump. AAPS needs to receive CGMS from other sources (BYODA or Xdrip are most common ways)

**Comment by jamorham** [comment]
xDrip should already work in companion mode with the G7 app. Let me know if not.

**Reply by steve8x8** [reply]
What about those without permanent internet connection, or a very slim data plan?

**Reply by Navid200** [reply]
You may be confusing the companion mode with web follower.
The companion mode does not need internet access.

**Reply by Navid200** [reply]
Assuming you have Dexcom app running on your phone (6 or 7), you have several options for passing the readings to xDrip:

1- Dex share, needs internet
2- Companion mode, does not need internet

When byoda comes out, we should test and confirm if that works with G7.  But, with G6, that's another option you have that does not need internet.

**Reply by steve8x8** [reply]
Thanks for clarifying - there was some confusion on my side what "G7 app" was referred to (and the regular one would not allow a "companion" to read from the device itself, right?). So everything for now boils down to either a BYODA-G7 or (or and) native xDrip access to the transmitter. 👍

**Reply by jamorham** [reply]
No, companion collector mode works with official apps.

**Reply by Miquest** [reply]
Cool, but my Smartphone is not compatible with the official app...
I will not be able to work with the App in Companion mode...

**Reply by AliQ101** [reply]
> No, companion collector mode works with official apps.
For me also G6 BYODA (G6 international v.1.10.0.2. with no direct broadcast do xDrip) works in companion mode.

**Reply by joelm335** [reply]
> No, companion collector mode works with official apps.

To activate companion mode, I assume I select Companion (Testing) as the data source.  I have a compatible phone although it is running Android 12, so not sure if it will work with G7 yet.

**Reply by Navid200** [reply]
@joelm335 Yes, that's how you activate it.  

As of the September 21, 2022 release, it does not show "Companion (Testing)" anymore.  
It shows "Companion App".  No change to functionality though.

**Reply by mistermintsgh** [reply]
So before I call my hospital team and ask to be switched across to G7, I just want to make absolutely sure I'm understanding what I'm reading, and it'll work with my setup....

I can install the official G7 app, no need to wait for G7 BYODA, and start sensors using that? Then in an xDrip release from later than 21/9/22 I select companion app as the BG source, and that won't require an internet connection? I can then feed the data from G7 app > xDrip > AAPS and complete the loop, which will then run my system pretty much the same as it does now with G6?

**Reply by Navid200** [reply]
> I can install the official G7 app, 

You can check on Dexcom site ( look for compatibility) to see if the G7 app can be installed on your phone or not.
If not, no, you cannot install the official G7 app.

If you can, xDrip will show the readings from the app on the same phone.  So, no need for internet.

**Reply by mistermintsgh** [reply]
It's showing as available on the Play Store, so regardless of that the Dexcom website says, or looks like it's compatible (Galaxy S21U)

As for the rest of my questions, that's all good? xDrip will get data from the official G7 app then feed that into AAPS?

**Reply by mistermintsgh** [reply]
Scrap that! The app will install, but once you log in it says the phone hasnt been tested and is therefore not compatible. You'd think they would just prevent it being visible to incompatible phones in the Play Store in the first place!

I guess I'm waiting for BYODA to be released

**Reply by Miquest** [reply]
I'm also ABLE to install the Dexcom App, but it sais always that I have no compatible device for it.

**Reply by Kitekater** [reply]
Take this one. My Samsung ist not compatible either. But with this G7 App, new xdrip with Companion Mode, i get it into AAPS.

How does this Companion Mode works?


https://github.com/DiaKEM/dexcom-g7-apk-patcher/releases/tag/v1.3.2.3326

**Reply by Miquest** [reply]
Wow, thank you so much, that helps a lot! :)

I'll try the G6 Version in a few days!

**Comment by blaqone** [comment]
iOS Loop has g7 Integration : https://github.com/LoopKit/LoopWorkspace/compare/dev...g7

Maybe work from there to get it going with xdrip!?

**Reply by jamorham** [reply]
This requires the official app installed like companion mode does.

**Reply by blaqone** [reply]
Oh okay. Then it’s useless.

**Comment by tim2000s** [comment]
@Navid200 let me know if you want a live or expired G7 to help with xDrip native implementation.

**Reply by blaqone** [reply]
The BYODA author needs a person with a g7. If you want to help?!

**Reply by Navid200** [reply]
Sent you a PM.

**Reply by tim2000s** [reply]
Which system?

**Reply by Navid200** [reply]
> Which system?

I have sent you a message on Gitter.

**Reply by Miquest** [reply]
Thank you so so much for your great engagement!

**Reply by Navid200** [reply]
@tim2000s I have sent you another message.  Thanks

**Reply by blaqone** [reply]
i have a few spare unused g7 I can provide you with.

**Comment by blaqone** [comment]
So im testing the G7 with Xdrip+. It works just find and gets al lthe data from the notifications to Xdrip. Somehow i disable a low bg warning tonight and then xdrip did not receive and data for a few hours. "Errorlog: Did not find any matches“ - Need to do a few more tests.

The meassured bg G6 vs G7 is pretty comparable within 10-15mg/dl. So thats a good thing.
 
BUT: Sadly the G7 App does not do any backsmoothing of the bg values it gets from the sensor. This looks like this:

![f0469d8003337c253e034e22676536a09ddd0877_2_897x748](https://user-images.githubusercontent.com/37814299/196028415-5ac2d6b4-365b-447c-968c-d198edf01d91.jpeg)
compared to the G6
![b67e9f547103241e525315f824e90bdfbb5ca044_2_1035x226](https://user-images.githubusercontent.com/37814299/196028417-913ab664-2569-4348-9996-7e47017a00b0.jpeg)

which makes it impossible to use in AAPS as a reliable bg source to loop with as the new received bg vary too much.

Another picture from the received G7 values within Xdrip:

![e4392aa93983cb1aacaeb32c85e87556dc32225e](https://user-images.githubusercontent.com/37814299/196028492-57f5d55e-9e8c-4aa9-9ea6-16fd5df0db96.jpeg)

Either Dexcom needs to update the G7 app (if they want to work with comercial loops) or Xdrip somehow enables a smoothing, compared to the L2/L3 (But the L2/L3 values come in every minute - G7 every 5.)

All in all - looping (AAPS) with the G7 is technically possible, in reality not feasible due to extreme fluctuations.

**Reply by jamorham** [reply]
Please can you confirm which xDrip version you are using and whether you are using the companion bluetooth feature? Thanks.

Smoothing is possible if we need to. Please also confirm if you are using xDrip graph smoothing feature already?

**Reply by blaqone** [reply]
Xdrip 2022.08.03
Im using "companion (Testing)"
Graph smoothing is anabled - let me know in the settings what exactly are you refereing to ?!

**Reply by blaqone** [reply]
Looks like i indeed used an old version - replaced with latest nightly.

**Reply by jamorham** [reply]
Please update to the latest nightly version and search for the `Companion Bluetooth` feature and try that. Disable it if there are any problems. It should enable back-filling in xDrip. There is also a feature `Graph Smoothing` try enabling that as well (but it wont affect AAPS)

**Reply by blaqone** [reply]
Perfect - did that! Will update my post above. Thank you!

**Reply by blaqone** [reply]
With the nightly build it works perfectly fine. The backfilled and smoothed graph is perfect for looping:

G6:
![fa515f19d92889e1b4df552264dc6958e9c600c5](https://user-images.githubusercontent.com/37814299/196032886-cd4cfff7-679b-42ea-9c50-7d0ecbd42a74.jpg)
and
![d56762f44fd3cd660e5519f9db300cddd81cd736](https://user-images.githubusercontent.com/37814299/196032904-25895a55-7c10-47e9-935e-e816b23e0bfa.jpg)


G7:
![57d2845696ef3ed727ea31ba3e5e14a78a81bde7](https://user-images.githubusercontent.com/37814299/196032915-3ab15124-5ec0-4709-b6d1-1028cf005d31.jpg)
![4425ab3bf0e4b19f0c93a62d956ff7f5bcf29730](https://user-images.githubusercontent.com/37814299/196032920-6086127c-d1d5-458a-9d0a-ec22db221175.jpg)

**Reply by Navid200** [reply]
As mentioned already, Graph smoothing does not affect the data.  It only affects the graph.

**Reply by blaqone** [reply]
As a result of the smoothing, the numbers sent to AAPS are „better“..

**Reply by Navid200** [reply]
> As a result of the smoothing, the numbers sent to AAPS are „better“..

@jamorham

**Reply by blaqone** [reply]
There are no more jumpy values coming from g7 over xdrip after enabling above settings.

**Reply by Kitekater** [reply]
Loop with AAPS works perfekt in this Combination. There are no spikes in the line even not in AAPS. Only in AAPS beside the sugar value, you have a yellow triangle with the notification" used date recalculated". SMBs are working fine. higher and lower basal is ok. Middle of my last three days testing the G7 was a value of 125 (target 120)

**Reply by blaqone** [reply]
It does not look good here. There is no smoothing and the values are all over the place:
![image](https://user-images.githubusercontent.com/37814299/196357624-1a9ca0cb-d3b8-40fa-bcee-b9a64734691b.jpg)

**Reply by blaqone** [reply]
Just used a 2nd sensor - same shit. 

> There are no more jumpy values coming from g7 over xdrip after enabling above settings.

This was incorrect. Xdrip companion mode read the data from the g6, not the g7.

**Reply by blaqone** [reply]
> Please update to the latest nightly version and search for the `Companion Bluetooth` feature and try that. Disable it if there are any problems. It should enable back-filling in xDrip. There is also a feature `Graph Smoothing` try enabling that as well (but it wont affect AAPS)

![59AF4A69-568B-4243-9BFD-8CFF65B74EE8](https://user-images.githubusercontent.com/37814299/197177915-08926592-9694-4b11-a05f-df116ef2e5cc.jpeg)
![0C3AE1EF-B472-48CD-8967-01E25B8C1003](https://user-images.githubusercontent.com/37814299/197177924-4817ae9e-9d19-4343-be11-15445a36be5e.jpeg)

So the sensors fills back data, but xdrip does not do that.

**Reply by blaqone** [reply]
Is there any way i can smooth the data BEFORE its being sent to AAPS? @jamorham

**Comment by blaqone** [comment]
The g7 itself is very unreliable. Had three sensor errors within the last 12 hours. Those are not pressure lows - they just happen…. Values are very jumpy too. 
![image](https://user-images.githubusercontent.com/37814299/196343415-4f6c0adb-dc33-47ae-bbc4-1be2a10b4796.jpg)

![image](https://user-images.githubusercontent.com/37814299/196343643-deee11c6-47b0-4060-9596-60440e339371.jpg)

**Reply by Miquest** [reply]
In my opinion, that are normal errors with a new product.
Maybe the next sensors will be better.

**Reply by blaqone** [reply]
Yeah I just replaced it. Overlapping warmup is nice thou.

**Reply by blaqone** [reply]
No change - new sensor same shit. 
![D14C666D-5CC7-4327-B1F5-EC77D5659A94](https://user-images.githubusercontent.com/37814299/196500006-533fc5fb-4fcd-4828-b670-69031240b277.jpeg)
![706F7D0B-2AD4-4A65-A090-CBC6D41171A0](https://user-images.githubusercontent.com/37814299/196500017-33e583ee-d0ee-437c-8538-45a0b0112d67.jpeg)

**Reply by blaqone** [reply]
The G7 does not backfill data when back from out of reach. WTF?

**Reply by Miquest** [reply]
In Germany we say "Kinderkrankheiten"...
I think, you have to wait for a Firmware Fix with the next Sensors..

**Reply by blaqone** [reply]
This is correct. The batch I am using was produced 01/05/22, so pretty early. I’ll receive a new batch tomorrow. We’ll see the production date…

**Reply by Miquest** [reply]
I hope the next batch will be better! 
I'm currently waiting for the G7 Update for my T:Slim X2. A few weeks or months (I cant remember) months before they announced that they're working on it as soon as the G7 is out.
Great that you tested it before, so that I can change calmly to the Dexcom G7.
Thank you!

**Comment by Navid200** [comment]
@blaqone Do you have a phone that is supported by the official Dexcom G7 app?

**Reply by blaqone** [reply]
Sadly no.

**Reply by Navid200** [reply]
How are you getting readings from G7?

**Reply by blaqone** [reply]
I tried the patched g7 app and the byoda version of it.

**Reply by tim2000s** [reply]
I'm wondering if this is an issue with the BYODA or Android  app. I've run the G7 on iOS and had both smoothing and very few issues with data consistency. I know they've struggled a bit with the Android roll out. 

I was also using a UK sensor, which is different to the European ones and were manufactured with the first gen algorithm on board, rather than the second gen.

**Reply by blaqone** [reply]
How can you tell that there are different algos?
I have run a few g7s on different phones - only Android thou - same jumpy readings on all.

**Reply by tim2000s** [reply]
The MARD evaluations between the UK and EU systems are different. 

Back when the first EMA approval was given, they announced that they would release in some countries with the originally approved algo and that an update would be coming later. 

As far as I can tell, the FDA accuracy study is the newer one, which is also what is in the EU sensors, as the MARD data matches, and the UK edition is the older one, as again, the MARD data matches. 

I might also have heard they'd had some issues with the Android app, which delayed the UK release. As I said, I've only used the iOS app as when I had a G7, the BYODA version didn't exist, and it was definitely not jumpy, and had no issues with smoothing, just like the old G6 apps. 

Will be using the EU version of the G7 shortly, on Android, so will provide update soon.

**Reply by blaqone** [reply]
Very interesting indeed!
Do you happen to have the sources handy? Would love to read those data studies.

**Reply by tim2000s** [reply]
I don't think they published the one that went with the EMA approval. It was just the numbers mentioned in articles at the time. 

Regarding the more recent algo, the details are here: https://pubmed.ncbi.nlm.nih.gov/35157505/

**Comment by adanvdo** [comment]
Read through this whole thing, but I'm still unsure. Is the official G7 app required in order to use xdrip+ with the G7?
There is no direct connection from xdrip right now? Are there plans for direct control or is it not possible?

**Reply by blaqone** [reply]
You can use the official G7 App or the patched one. Both works with xdrip in companion mode. 

The patched g7 App can also send the data directly to AAPS.

**Reply by tim2000s** [reply]
Are you sure about that last point? It's not stated on the form, and I'm not aware that anyone has tested it.

**Reply by blaqone** [reply]
Yes, I am very sure, as I am using it for a few weeks now. I worked with @khskekec and tested it for quite some time now.

The patched app can be found here: https://github.com/DiaKEM/dexcom-g7-apk-patcher

And this is the latest master AAPS with G7 support, fixed a few bugs, added autoIFS and g7 value smoothing : https://github.com/DiaKEM/dexcom-g7-aaps/tree/blaqone-final

**Reply by tim2000s** [reply]
Okay, that's not traditionally what people consider to be the patched app. Pleased to hear it works though. 

The traditional use of that term refers to the BYODA app, accessed via the Google form.

**Reply by adanvdo** [reply]
I dont use AAPS, I don't even know what it is. I currently use Nightscout with the G6. I've also never used companion mode.
So, based on what you are saying, xdrip will not handle G7's the same way it currently handles the G6.

**Comment by Chris99621** [comment]
I use the G7 with the companion mode too. Sadly the battery drain from the G7 App is very high. A direct implementation in xDrip would be nice for the battery. The G7 App uses more Power than AAPS. With the Libre before i never noticed a battery drain from the Libre App. I Hope someone will implement the G7 in xDrip soon.

**Reply by xannasavin** [reply]
I'm not yet on the G7 but I would also appreciate a lot, if the support was build into xdrip rather than xdrip being a trailer on the native app. Xdrip is perfectly reliable (using it since several years now) and two apps for the same purpose doesn't make much sense, also in terms of battery life.

**Reply by TobiasWegener** [reply]
Are there any news on this topic, I just got my first batch of G7 from Sensors produced on 06.2022 and now I am looking to how to read them out. Can I do anything, to increase the chance for the development of native xDrip support for G7?

**Reply by blaqone** [reply]
Right now you either use the patched g7 app to send values directly to AAPS or use xdrip’s companion mode.

**Reply by gitanO0** [reply]
Shucks.  I've been using only xdrip+ and g6 (no pump/loop) for years and would be terrible if the g7 severs this functionality.  Is it a sure thing that xdrip+ and g7 can't work together as xdrip+ and g6 do?  I really don't want to install the native Dexcom/g7 app on my phone AND run xdrip+ on top of that.  Booooooooo.

**Comment by Arakon** [comment]
I'm running Xdrip+ and the latest G7 app released today. It no longer activates GPS, so that should reduce the battery drain.. however, I am not getting any data from it. I have both Companion App and Companion Bluetooth active.
I am currently running a G6 and G7 in parallel. Is it normal for Xdrip+ to ignore the G7 data in that case (they're ~2 minutes apart)? The new quick glance notification of the G7 app looks slightly different from before, in case that may make a difference.

![image](https://user-images.githubusercontent.com/12016360/218864956-4c5aa59e-9bb3-4bf4-83dd-0a501ed98aa5.png)

**Reply by blaqone** [reply]
You can not run two sensors and grab data with xdrip in companion mode. Either or… I had the same setup a few months ago. Same result.

**Reply by Arakon** [reply]
> You can not run two sensors and grab data with xdrip in companion mode. Either or… I had the same setup a few months ago. Same result.

Alright, thanks. Was just surprised that it basically completely ignored the existance of the second sensor with no error in the log.

**Reply by Navid200** [reply]
Disable the notifications from one app and enable them from the other.  xDrip will continue to show them from the app that provides notifications.

Let us know if it doesn't.

**Reply by Arakon** [reply]
> Disable the notifications from one app and enable them from the other. xDrip will continue to show them from the app that provides notifications.
> 
> Let us know if it doesn't.

Indeed, you are right.. just had to also disable Companion Bluetooth or it would still continue to receive the data from the G6. After that, it switched over to the G7 data stream. So yes, it still works with the new app version.

**Reply by blaqone** [reply]
Of course. Just noteworthy both at the same time.

**Comment by wadert3** [comment]
So from reading it sounds like theres not going to be g7 compatibility with xdrip+? Was hoping to see if there would be an update and a possibility of sometime in the near future having it connect directly with the g7 rather than having to run both apps. Going back and forth on whether I should move to the g7 or not.

**Reply by Unrealsweb** [reply]
Das wäre sehr schade, es geht nichts über only Xdrip

**Comment by xannasavin** [comment]
What are the current obstacles regarding the G7 support? Why is it not possible to integrate it like for G5/G6? Is it only a lack of manpower or also some serious issues with communication, code, whatsoever?

**Reply by yoshimo** [reply]
On the IOS Side there is work https://github.com/gui-dos/DiaBLE/commits/main for Dexcom Sensors so i think it comes down to manpower. Usually with XDrip it is a lack of a capable person that has the right devices and is intrested in making it work.

**Reply by wadert3** [reply]
How does the android group look? Is there a viable group on a discord or something that is working on it? I may be able and willing to assist if possible depending on who/ what's being done currently

**Comment by Ferdinandoz** [comment]
Does anybody know whether it will be possible in the future to use xdrip with the G7 without having another Dexcom app installled/running?

So is it a question of time or is it impossible? This is maybe important for many users who are going back and forth whether changing from G6 to G7.

**Reply by Navid200** [reply]
Don't delay any decisions because someone may tell you that yes, xDrip will have it.

xDrip may have the feature.  We all hope that it will.
But, there is no way we can make a promise.  I cannot.  Such a promise does not help anyone.

Go ahead and do what you need to do to control your, or your loved one's, diabetes with what you have available to you today.  
If/when xDrip adds the feature, you can switch.

**Reply by dobey** [reply]
I found [this page](https://navid200.github.io/xDrip/docs/Dexcom/G7.html) when searching for G7 support in xdrip. Does this not mean that the G7 can be used as if it were a G6 sensor, @Navid200 ?

**Reply by Navid200** [reply]
@dobey  You can give it a try as explained here as well:  https://github.com/NightscoutFoundation/xDrip/discussions/2778

**Reply by TobiasWegener** [reply]
So, I got xDrip+ working with G7, which is fantastic, thank all the developers for the hard work!
In the end, I removed xDrip and restart phone and then installed xDrip again and followed exactly the steps in [#2778](https://github.com/NightscoutFoundation/xDrip/discussions/2778).  
Complication for me were the old Sensors still sending data and fighting for prevalence (I finally put them all in a metal box) and still having Dexcom app installed. Solving both of these issues, restarting and reinstalling, seemed to do the trick.

Can confirm that back loading old data works, too. :-)

I will do calibration measurement soon and report back.
![Screenshot_20230419-172555_xDrip+](https://user-images.githubusercontent.com/31315723/233125023-9ea4b8cb-079b-4f80-9e7f-eb373efa8186.png)

**Reply by TobiasWegener** [reply]
So Calibration is definetly needed it was about 80 mg/dl higher then my accu check guide stated (red squares).
![Screenshot_20230419-204551_xDrip+](https://user-images.githubusercontent.com/31315723/233171289-2638fd3c-e4df-4f35-8847-86a308f9bcec.png)

Connection reliability seems much bettter, then with the original Dexcom APP.

**Reply by TobiasWegener** [reply]
Calibration helped and the dexcom 7 values fit mich better to my Accu check guide values:
![Screenshot_20230421-134726_xDrip+](https://user-images.githubusercontent.com/31315723/233628416-b724a6bd-ba96-48c6-a9ac-eb37fc5581ca.png)

**Comment by grahamjenson** [comment]
I am wondering if anyone else is seeing the issues with G7 in companion mode where sometimes it misses readings or gives the same reading twice. And it seems to get out of sync and get behind by minutes.

I used xDrip in companion mode with CamAPS before and it did not have these issues. I am thinking it might be an issue with the G7 App sending multiple notifications, or something. I am a bit familiar with the UIBasedCollector code, if there is anything I can do to help please let me know.

**Reply by Navid200** [reply]
You may want to open a new thread.  This thread is titled G7 support.  G7 support has now been added to xDrip.

**Reply by grahamjenson** [reply]
https://github.com/NightscoutFoundation/xDrip/discussions/2803

**Comment by Unrealsweb** [comment]
Will Xdrip still work with the G6 sensor and transmitter after the patch?

**Reply by Navid200** [reply]
What patch?
There has been no change that could impede the operation of xDrip with G6?

**Comment by Unrealsweb** [comment]
![g7pic](https://user-images.githubusercontent.com/6748067/235339650-d84e29c7-c28f-40f2-a563-7c40e021f440.jpg)


Calibration Code from G6 under Systemstatus

**Reply by Navid200** [reply]
The obvious problem, shown in that screenshot, is that xDrip does not recognize your device.
It will after this PR is merged: https://github.com/NightscoutFoundation/xDrip/pull/2792

If the device is recognized and a G6 calibration code from the past is still shown, we can fix it.

**Comment by Unrealsweb** [comment]
![image](https://user-images.githubusercontent.com/6748067/235907692-879c94ec-62d0-4e72-b067-5a7b1773313d.png)
![image](https://user-images.githubusercontent.com/6748067/235908282-e8993073-e6d6-4c0a-8837-130925e6b2ad.png)
Bluetooth connection lost and can no longer be connected

**Reply by Navid200** [reply]
@Unrealsweb 
Please update to the latest Nightly (https://github.com/NightscoutFoundation/xDrip/releases/tag/2023.05.03).
Please tell me if the problems persist.

**Reply by Unrealsweb** [reply]
The problem no longer exists, it was due to too many old G6 transmitters in the apartment
![image](https://user-images.githubusercontent.com/6748067/236146387-a0295551-1973-4917-998b-f6a201cd27ad.png)

**Reply by Unrealsweb** [reply]
Calibration Code is from old G6 Transmitter

**Reply by Navid200** [reply]
Alright
You can ignore the G6 calibration code until we remove it.  Thanks for reporting it.

**Comment by Unrealsweb** [comment]
Can't wear two sensors with Xdrip :( so activate one and set the other 6/h before and then activate it.

As soon as the second sensor has been set, the following happens.
![image](https://github.com/NightscoutFoundation/xDrip/assets/6748067/15a0eb62-6c43-4b96-8cfc-cdfcb6b887d0)
![image](https://github.com/NightscoutFoundation/xDrip/assets/6748067/7c813ae8-9eaf-42b2-98ce-fe1c75628581)

**Reply by adanvdo** [reply]
@Unrealsweb this isnt related to general G7 support. I would open another discussion.  However, I question why in the world you are trying to wear multiple sensors in the first place?  That is not how the product was intended to be used.

**Reply by Unrealsweb** [reply]
The original app has no problems with it.

And Xdrip does not even know the second sensor, the ID is not entered.
This would also happen if two users with Xdrip and G7 were within 6 meters.

**Reply by adanvdo** [reply]
What is the point of wearing 2 sensors though?

**Reply by Arakon** [reply]
Wearing 2 sensors at once is pretty common with the G7. You place a new sensor (but don't pair it) when the old one expires, then let the old sensor run the 12h grace period before switching over to the new sensor, which will then have had 12h of time to settle and give decent values. That way you lose no uptime at all.

**Reply by adanvdo** [reply]
While that might make sense, it didnt appear to be what they were trying to do. They were trying to start two overlapping sensor sessions.

**Reply by Navid200** [reply]
@Arakon This is being addressed here:  https://github.com/NightscoutFoundation/xDrip/discussions/2828

I asked these discussions to be opened.  The only objection to my suggestion was that it will get messy and you will need to spend even more time keeping things organized.
So far, I am pleased with these discussions having been opened.  However, I need the user community to support me.

Would you please be so kind to help me keep the discussions coordinated?
This discussion you are posting in was never meant to become about wearing multiple sensors.

There is absolutely nothing wrong with having a discussion about that or many other subjects.
But, this is a discussion that was opened with a particular purpose.  It is now turning into a completely different subject.

**Reply by adanvdo** [reply]
apologies @Navid200

**Comment by asavageiv** [comment]
I was to give a huge THANK YOU to the people that implemented G7 support in xDrip+. It really saved my ass.

I am from the USA and was in Mexico when my phone died. I bought a new one, but couldn't log in to the Dexcom G7 app. It's not allowed to log in to the app unless your phone is physically in the USA (or whatever country matches your Dexcom account), an absurd restriction.

If it wasn't for this support in xDrip I would have had to fly back to the USA over a month early _just to log in to the app_ or go back to finger sticks. So, once again: THANK YOU!

**Comment by jobermayr** [comment]
@Navid200 Is there a chance to get [your required QR code for G7](https://navid200.github.io/xDrip/docs/Dexcom/G7.html) integrated in xDrip to make the G7 work out of the box?

Interestingly I could make a sensor working with xDrip where binary blob doesn't get/show data anymore :)

**Reply by Navid200** [reply]
> Is there a chance to get [your required QR code for G7](https://navid200.github.io/xDrip/docs/Dexcom/G7.html) integrated in xDrip to make the G7 work out of the box? 

It shouldn't be such an inconvenience to scan that.  Is it?  Otherwise, may I ask the reason?

**Reply by jobermayr** [reply]
You need a second device to show it and I couldn't find a way to scan the PNG directly with only one device available.
Is there something that cannot be shipped with xDrip?

**Reply by Navid200** [reply]
I never thought of it that way.
You can print it on a piece of paper and scan that.  But, you only need to do that once.  You don't need to do it for every transmitter.

There is too much in xDrip.  Anything that can be kept out, I vote for that.
The following is relevant but humorous.   Nothing offensive.  I hope not.
https://youtu.be/8ZvugebaT6Q

Now, you may wonder "why did I just watch that"?!
That's xDrip.    Everyday, someone says "would you please add this additional feature"?
And I keep screaming NOOOOOOOOO!  But, no one hears me!
Did you see the guy who kept opening the door and said "... just the woman I'm looking for" and let her in?!  That's the developers.  They keep accepting to add more and more and  ....
Did you see what happened at the end?  I fear that's what's going to happen to xDrip sooner or later.

**Reply by jobermayr** [reply]
I know how much work it is to keep code working on new revisions (https://github.com/jobermayr/ghidra-staging).

I also understand your concerns but xDrip should be able to drop code for old devices like FSL2 which aren't available anymore sometime soon.
So maybe the code can lighten up somewhere else :)

It should be also easier to debug as developer/maintainer if you know exactly which version (and modifications) somebody has and don't need to collect modifications from different places (https://tsdgeos.blogspot.com/2010/05/complex-systems-are-complex.html?m=1).

So I am voting to push as much as possible to upstream to get it maintained and tested on changes as well as to make the entrance for new users easier.

**Reply by ah7675** [reply]
Nowadays, FSL1 is still available for Japan and China. I believe FSL2 will still available for years.
There are more pumps get the FDA clearance which can integrate FSL2 as loop.

**Reply by MisterMcDuck** [reply]
I don't want to bite the hand that feeds and am really happy/excited for the G7 support (having read through OB1/android BT code before) that appears to have been done by @Navid200, but with that said I think it's a bad idea to have support for what is now (in America at least) the main supported Dexcom CGM device implemented as what seems to be a plugin system. 
I can understand the complexity concerns, but what happens when we reach Dexcom G12+, and even more CGM devices are hopefully supported as well? Is it going to be a web of disparate githubs pulling support for each device?

FWIW I maintain my own version with minor changes and compile xDrip+, and having this external dependency makes it difficult to operate. 

Again, thanks so much for the contributions. I really respect anyone that contributes their time towards this essential application.

**Reply by asavageiv** [reply]
That link is broken Navid200. FWIW, as a professional software engineer myself, although I understand the desire to minimize the amount of code in the app, I think G7 support should be integrated into xDrip+ as a first class citizen. It could really help people in dire circumstances.

Let me explain:
I was in Mexico when my phone died. I got a new phone but I couldn't log into Dexcom because they geo-fence app login to the USA for some stupid reason. I then went to xDrip+ for support. Initially it didn't work at all. Luckily, I stumbled across the QR code, but someone less determined than me easily could have given up.

I assume G7 support will have to be integrated eventually anyway because G5 and G6 will disappear from the world. If you can link me to the source for the QR code and existing support I could take a look at it as well as looking at how to make such changes easier within the xDrip+ architecture.

There will always be new devices that should be added to xDrip+, so we should make it as easy as possible to avoid the problem you pointed out in your video.

**Reply by Navid200** [reply]
Thanks for letting me know that the link I provided does not exist anymore.  I am not sure what happened to it.  But, I suspect that it has been deleted.

Believe me, we know very well how to incorporate what's in a QR code.

xDrip is a tool that was created and is maintained by those who have chosen to use it.
Sharing it on GitHub has only been done so that individuals who understand that they have to choose to use it if they want to and do not blame those who have chosen to share it for any unexpected outcome.
This is a very different scenario than someone creating an app and sharing it and using advertising to make money or directly charging customers to offer a professional version of the app.
We do none of that.  There is absolutely no financial incentive for any of us tied to the use of xDrip.

I have diabetes.  I choose to use xDrip.  Over the years, I have gradually also started to contribute to xDrip as a junior developer as well as supporting users.
If that is going to result in some day being prosecuted, it makes no sense and will not be fair.  I perfectly understand that xDrip is not a medical app and should not be used as such.  No one can use xDrip before acknowledging that.

If anyone cannot use xDrip as is, there is no problem for them to stop using it.
If someone chooses to use xDrip, it does not give us any financial benefit.

All feedback is taken into account.

**Reply by MisterMcDuck** [reply]
I can only speak for myself, but part of the reason I participated in this thread was I had a critical misunderstanding of the current state of the G7 logic. I thought, due to earlier communications and initial development, that scanning the QR code above resulted in xDrip+ importing a compiled 3rd party library not represented in the xDrip github. 
My current understanding, thanks to other contributors, is that is **not** the case. The library logic (libkeks, etc) has now been brought into the xDrip project. 
I now understand that the QR code is purely a signed certificate which is used to interact with any G7 transmitters that apparently now need a signed, valid certificate. This seems to be left configurable for any of the users in case the currently used certificate expires or is revoked by Dexcom. 
If my current understanding is correct, I don't have any issues with the current implementation and again thank @Navid200 and others for their work on xDrip.

**Comment by MisterMcDuck** [comment]
Is the source code available for these changes? If so, where?

**Comment by jamorham** [comment]
This thread dates back to 2022 and really doesn't serve any purpose now as there has been full support for a long time. I will close it as every comment causes everyone who previously has commented to get pinged with a notification.
